File: flattennestedlistiterator.js
//Objective is to design a class for a nested list iterator.


//Design of the class using a dfs traversal to get all elements in a 1D array.

class NestedListIterator {
    constructor(list) {
        this.arr = []
        this.position = 0

        let dfs = (list) => {
            for (let element of list) {
                if (isFinite(element)) {
                    this.arr.push(element)
                } else {
                    dfs(element)
                }
            }
        }
        dfs(list)
    }

    hasNext() {
        return this.position < this.arr.length
    }

    next() {
        return this.arr[this.position++]
    }
}

let list = new NestedListIterator([[1,1],2,[1,1]])
list.next()
list.next()
list.next()
list.next()
list.next()

File: flattennestedlistiterator.java
//Java Solution

public class NestedIterator implements Iterator<Integer> {

    int index;
    List<Integer> arr;
    public NestedIterator(List<NestedInteger> nestedList) {
        index = 0;
        arr = new ArrayList<>();
        dfs(nestedList);
    }
    
    public void dfs(List<NestedInteger> nestedList) {
        for (NestedInteger element : nestedList) {
            if (element.isInteger()) {
                arr.add(element.getInteger());
            } else {
                dfs(element.getList());
            }
        }
    }

    @Override
    public Integer next() {
        return arr.get(index++);
    }

    @Override
    public boolean hasNext() {
        return arr.size() > index;
    }
}